Add sequential workshop tutorial format (pick-and-place seed) (#5104)
viamrobotics/docs: docs/tutorials/pick-and-place/02-configure-resources.md

linkTitle: "2. Configure resources"
type: "docs"
slug: "configure-resources"
weight: 20
description: "Configure the arm, gripper, and camera by hand, then verify each one from the control and 3D scene tabs."
workshop: "pick-and-place"
toc_hide: true
phase: 2
phase_total: 6
time_estimate: "20 minutes"
prev: "/tutorials/pick-and-place/platform-mental-model/"
next: "/tutorials/pick-and-place/static-positions/"
languages: ["python"]
---

In this phase you configure the three hardware resources every pick-and-place machine needs: an arm, a gripper, and a camera. Your machine starts with none of them. You add each one by hand, watch it come online, and confirm it works from the CONTROL tab before moving to the next.

{{< workshop-phases >}}

## The target state

By the end of this phase your CONFIGURE tab holds three components:

| Name        | API                     | Model                   |
| ----------- | ----------------------- | ----------------------- |
| `arm-1`     | `rdk:component:arm`     | `viam:ufactory:xArm6`   |
| `gripper-1` | `rdk:component:gripper` | `viam:ufactory:gripper` |
| `cam-1`     | `rdk:component:camera`  | `viam:camera:realsense` |

> You will not touch the vision service in this phase. The `shape-detector` and `vision-segment` services that let the machine find blocks by shape come later, in Phase 5, right before you write the code that calls them. For now, focus on the hardware.

Reviewer: ```suggestion

```

## Configure the arm

<!-- ASSET P0 configure-add-component (UI+): add-component dialog, "xArm6" searched, viam:ufactory:xArm6 result highlighted. See plans/2026-07-02-pick-and-place-shot-list.md -->
<!-- ASSET P0 logs-xarm-module-start (UI+): LOGS showing the viam:ufactory module download + start (the module-download moment) -->
<!-- ASSET P1 configure-arm-attributes (UI): arm-1 card with host / port 502 -->

> On the **CONFIGURE** tab, add a new component. In the add-component dialog, search for `xArm6` and select the `viam:ufactory:xArm6` result. Name the component `arm-1`.

Reviewer: they renamed them "blocks" recently, + > Blocks -> search

> On the **CONFIGURE** tab, add a new component. In the add-component dialog, search for `xArm6` and select the `viam:ufactory:xArm6` result. Name the component `arm-1`.

Reviewer: ```suggestion
On the **CONFIGURE** tab, add a new component. In the add-component dialog, search for `xArm6` and select the `ufactory/xArm6` result and press Add to machine. Name the component `arm-1`.
```

Set the following attributes:

- `host`: the xArm controller's IP address, provided by your workshop facilitator
- `port`: `502`

You can leave `speed_degs_per_sec` and `acceleration_degs_per_sec_per_sec` unset for now; both are optional and default to safe values.

> Save the config. This is the moment from Phase 1 made concrete: `viam-server` does not have `viam:ufactory:xArm6` built in, so it fetches the `viam:ufactory` module from the Viam registry and starts it. Open the **LOGS** tab and watch it happen: a log line for the module download, then one for the module starting, then `arm-1` itself coming online. The whole sequence usually takes well under a minute.

Reviewer: the "made concrete" sounds funny, but better phrasing is escaping me

{{< alert title="Two components, one module" color="note" >}}

> The `viam:ufactory` module provides both the arm model and the gripper model you configure next. You only pay the download cost once.

Reviewer: ```suggestion
The `viam:ufactory` module provides both the arm model and the gripper model you configure next. Viam Server only has to download the module once to access both models.
```

{{< /alert >}}

## Configure the gripper

<!-- ASSET P2 configure-gripper (UI): gripper-1 config with arm: "arm-1" -->

> Add another component. In the add-component dialog, search for the `viam:ufactory:gripper` model and select it. This model comes from the same `viam:ufactory` module as the arm. Name it `gripper-1`.

Reviewer: ```suggestion
Add another component. In the add-component dialog, search for the `ufactory/gripper` model and select it. This model comes from the same `viam:ufactory` module as the arm. Name it `gripper-1`.
```

Set one attribute:

- `arm`: `"arm-1"`

> 

Reviewer: i think we need to add a frame connecting the gripper to the arm

This attribute is also a dependency: `gripper-1` cannot start until `arm-1` is running, because the gripper is physically mounted on the arm and controlled through the same connection. Save the config and check the **LOGS** tab again. Because `viam-server` already has the `viam:ufactory` module running from the previous step, `gripper-1` comes online immediately with no second download.

## Configure the camera

<!-- ASSET P2 configure-camera (UI): cam-1 realsense config (sensors color+depth, 640x480) -->

> Add a third component. In the add-component dialog, search for `realsense` and select the `viam:camera:realsense` result. Name it `cam-1`.

Reviewer: ```suggestion
Add a third component. In the add-component dialog, search for `realsense` and select the `realsense/realsesne` result. Name it `cam-1`.
```

Set the following attributes:

> 

Reviewer: does this need the SN? or does it just grab the first one it finds if non exist?

- `sensors`: `["color", "depth"]`
- `width_px`: `640`
- `height_px`: `480`

> Save the config and confirm in the **LOGS** tab that `viam-server` downloads the `viam:camera-realsense` module and starts `cam-1`. This is a separate module from `viam:ufactory`, since it comes from a different publisher and family.

Reviewer: i think we need to add a frame connecting the camera to the arm

## Test each resource from the CONTROL tab

<!-- ASSET P1 control-camera-stream (UI): CONTROL camera card, live color + depth -->
<!-- ASSET P1 control-arm-card (UI+): joint sliders + Get end position boxed -->
<!-- ASSET P1 control-gripper-grab (MOTION): gripper closing on a block via Grab, then Open releasing -->

Open the **CONTROL** tab. You should now see a test card for each of the three components you just added.

On the camera card, confirm you get a live feed:

{{< checkpoint >}}

> The camera card shows a live color stream from `cam-1`. If depth is available as a separate view, confirm that stream updates too. If the card is blank, check the LOGS tab for a camera error before moving on.

Reviewer: the depth should be there? can we word it differently?

{{< /checkpoint >}}

> On the arm card, jog a joint with the sliders and confirm the arm moves. Then select **Get end position** and confirm it returns x, y, and z coordinates.

Reviewer: I think this needs a warning that the arm is about to move and we need to check the estop, and i think joint sliders can cause collisions

{{< checkpoint >}}
Jogging a joint slider moves the physical arm, and **Get end position** returns a coordinate rather than an error. If jogging does nothing, confirm `arm-1` shows as online in the CONFIGURE tab and that the LOGS tab has no connection errors for it.
{{< /checkpoint >}}

On the gripper card, place a block between the gripper fingers by hand, then select **Grab**. The fingers should close and hold the block. Select **Open** and confirm the fingers release it.

{{< checkpoint >}}
With a block between the fingers, **Grab** closes the fingers and the gripper holds the block without dropping it. **Open** releases the block. This grab-and-release is the same action your Python code performs later in the workshop when it picks a block and drops it in a bin. If your gripper card also shows a holding status indicator, it now reads true while the block is held and false once the gripper is open and empty.
{{< /checkpoint >}}

> ## The 3D scene tab

Reviewer: I think we skipped the frame part for these to connect to eachother

<!-- ASSET P0 3dscene-wrist-camera (MOTION): jog joint 1 and watch the cam-1 frame move with the arm (the wrist-camera insight) -->

Open the **3D scene** tab. You should see a rendered model of the arm and the `cam-1` frame positioned at the end of its wrist.

Jog joint 1 with the arm card's sliders and watch the 3D scene as the arm turns. The `cam-1` frame moves with the arm, because the camera is mounted on the wrist rather than fixed in the workspace.

{{< alert title="Why this matters later" color="note" >}}

> Because the camera is wrist-mounted, every detection it makes is relative to wherever the arm happens to be pointed at that moment. In Phase 5 you will detect from a single known pose so that "camera space" always means the same thing. Notice that pose here; you will meet it again as the "detect from home" rule.

Reviewer: i think detection and camera space feel weird and might need to replaced with something slightly more clear for the reader?

{{< /alert >}}

> The exact camera mounting offset used to transform its frame into the arm's frame comes from your hardware setup guide or the pre-provisioned config, not from anything you configure in this phase. For now, just confirm the relationship: the frame exists, and it tracks the arm.

Reviewer: maybe consider moving this to later in perception guided picking
?

## Check your work

> 

Reviewer: maybe list what should be in the config and what should be seen in the 3d view before referencing the json. might be able to just render it here but pull from github so it always stays up to date

If you want to compare your configuration against a known-good version, the companion repo has a reference copy at [machine-fragment.json](https://github.com/viam-devrel/pick-and-place/blob/main/config/machine-fragment.json). Use it to check your work, not to import over what you just built by hand.

> With `arm-1`, `gripper-1`, and `cam-1` all live and verified, you are ready for Phase 3, where you save a set of fixed arm poses and prove the full hardware sequence works before perception enters the picture.

Reviewer: ```suggestion
With `arm-1`, `gripper-1`, and `cam-1` components all live and verified, you are ready for [Phase 3, where you move the arm through a set of arm poses](03-static-positions.md).
```

<!-- ASSET P0 discovery-test-panel (UI+): realsense discovery TEST panel showing a camera config snippet with serial_number filled in -->

{{<imgproc src="/tutorials/pick-and-place/discovery-test-panel.png" resize="1200x" declaredimensions=true alt="The RealSense discovery service TEST panel showing a discovered camera config with its serial number.">}}

The discovered component arrives named `realsense-<serial_number>`, with `sensors` set to `["color", "depth"]` and `serial_number` already filled in. Rename it to `cam-1` so it matches the rest of this workshop, then add the one attribute discovery does not set:

- `align_color_depth`: `true`

Leave the discovered `sensors` and `serial_number` as they are; letting discovery set the serial number is the whole point of using the service. You do not need to set `width_px` or `height_px`; the module uses a supported default resolution, and pinning one that the camera cannot produce would fail the build.

<!-- ASSET P2 configure-camera (UI): cam-1 realsense config, sensors color+depth, align_color_depth true, serial_number populated -->

> {{<imgproc src="/tutorials/pick-and-place/configure-camera.png" resize="1200x" declaredimensions=true alt="The cam-1 config with color and depth sensors, align_color_depth enabled, and a serial number.">}}

Reviewer: Might be good to red box the camera name and the attributes to make sure the camera rename isn't missed

> `align_color_depth` is the attribute that makes perception work later. With it set to `true`, the module aligns each depth frame to the color frame, so a given pixel in the color image and the same pixel in the depth image describe the same physical point. The `vision-segment` service in Phase 5 relies on that alignment to turn a 2D detection into a 3D point cloud segment. Both `color` and `depth` must be in the `sensors` list for it to take effect.

Reviewer: ```suggestion
`align_color_depth` When set set to `true`, the module aligns each depth frame to the color frame, so a given pixel in the color image and the same pixel in the depth image describe the same physical point. The `vision-segment` service in Phase 5 relies on that alignment to turn a 2D detection into a 3D point cloud segment. Both `color` and `depth` must be in the `sensors` list for it to take effect.
```

Save the config and confirm in the **LOGS** tab that `cam-1` starts. Because `viam-server` already downloaded the `viam:realsense` module when you added the discovery service, the camera comes online with no second download.

{{< alert title="The discovery service has done its job" color="note" >}}
The discovery service is not part of the pick-and-place pipeline; it only helped you configure `cam-1`. Leave it in place to re-discover hardware later, or remove it once the camera works. The rest of the workshop does not use it.
{{< /alert >}}

## Connect the components with frames

> Adding the arm, gripper, and camera tells `viam-server` how to talk to each one, but not where each sits in space. The motion service needs that spatial relationship: it has to know the gripper rides on the end of the arm and the camera looks out from the wrist, or it cannot plan a pick. You supply it by adding a **frame** to each component.

Reviewer: ```suggestion
Adding the arm, gripper, and camera tells `viam-server` how to talk to each one, but not where each sits in space. The motion service needs that spatial relationship: it has to know the gripper is mounted to the end of the arm and the camera looks out from the wrist, or it cannot plan a pick. You supply it by adding a **frame** to each component.
```

The three frames form a tree rooted at `world`:

```text
world
└── arm-1
    ├── gripper-1   (rides on the end effector)
    └── cam-1       (mounted on the wrist)
```

### Frame the arm

> Open the `arm-1` card and select **Frame**. The default frame already describes what you want: parent `world`, translation `(0, 0, 0)`, no rotation. That places the arm's base at the origin of the world, and every other frame is measured from there. Leave the defaults and save.

Reviewer: ```suggestion
Open the `arm-1` card in the configure tab and select **Frame**. The default frame already describes what you want: parent `world`, translation `(0, 0, 0)`, no rotation. That places the arm's base at the origin of the world, and every other frame is measured from there. Leave the defaults and save.
```

<!-- ASSET P1 configure-arm-frame (UI): arm-1 card Frame editor, parent world, translation 0,0,0 -->

{{<imgproc src="/tutorials/pick-and-place/configure-arm-frame.png" resize="1200x" declaredimensions=true alt="The Frame editor for arm-1 with parent world and zero translation.">}}

### Frame the gripper

Open the `gripper-1` card and select **Frame**. Switch the editor to JSON and set:

```json
{
  "parent": "arm-1",
  "translation": { "x": 0, "y": 0, "z": 105 },
  "orientation": {
    "type": "ov_degrees",
    "value": { "x": 0, "y": 0, "z": 1, "th": 0 }
  }
}
```

> Parent `arm-1` attaches the gripper to the end of the arm, and `z: 105` places the gripper frame 105 mm out from the arm's end effector, at the gripper's tool center point. This is the point the motion service drives to a target when you call `motion.move` later in the workshop, and the reference the grasp math in Phase 5 builds on. This frame is separate from the `arm` attribute you set earlier: the attribute shares the arm's connection, while the frame tells the planner where the gripper sits.

Reviewer: this introduces the tool center point,, I think this is the first time? Might be good to make a diagram or grab a photo of the real arm.  It also introduces parent frames, might be good to add a few extra sentences or link to the docs and let readers choose their own adventure

### Frame the camera

Open the `cam-1` card and select **Frame**. Switch the editor to JSON and set:

```json
{
  "parent": "arm-1",
  "translation": { "x": -73, "y": 40, "z": 18 },
  "orientation": {
    "type": "ov_degrees",
    "value": { "x": 0, "y": 0, "z": 1, "th": 270 }
  }
}
```

> Parent `arm-1` mounts the camera on the wrist so it moves with the arm. The translation places the camera's optical center relative to the end effector, and the 270-degree rotation around z lines the camera's axes up with the arm's. Save the config.

Reviewer: ```suggestion
The camera mounts to the same mounting flange the hand does so it has the same parent frame: `arm-1`. You have to tell Viam where the `optical center` is, the place on the camera the camera and point cloud are expressed in. The translation places the camera's optical center at the focal point of the camera imager with respect to the arm flange, and the 270-degree rotation around z lines the camera's axes up with the arm's. Save the config.

Check that your camera configuration matches the image below: Check the component name is cam-1, the module / model selected is realsense/realsesne and the frame values match. Note that the attributes may change their order, you only need to verify the values,
```

<!-- ASSET P0 configure-camera-frame (UI+): cam-1 Frame editor JSON, parent arm-1, translation -73,40,18 -->

{{<imgproc src="/tutorials/pick-and-place/configure-camera-frame.png" resize="1200x" declaredimensions=true alt="The Frame editor JSON for cam-1 with parent arm-1 and a wrist-mount translation.">}}

{{< alert title="These values match this workshop's hardware" color="note" >}}
The translation and orientation above describe the specific gripper and camera mount used in this workshop. If your hardware differs, work through the [frame calibration worksheet](https://github.com/viam-devrel/pick-and-place/blob/main/setup/frame-calibration-worksheet.md) to measure your own offsets, and see [Configure frames for an arm, gripper, and wrist camera](/motion-planning/frame-system/arm-gripper-camera/) for the full method.
{{< /alert >}}

> ## Test each resource from the CONTROL tab

Reviewer: the 3 sub sections could maybe use title text, Test The Camera, Move the Arm, Control the Gripper

Open the **CONTROL** tab. You should now see a test card for each of the three components you just added.

On the camera card, confirm you get a live feed:

<!-- ASSET P1 control-camera-stream (UI): CONTROL camera card, live color + depth -->

{{<imgproc src="/tutorials/pick-and-place/control-camera-stream.png" resize="1200x" declaredimensions=true alt="The cam-1 CONTROL card showing a live color stream of the blocks.">}}

> {{< checkpoint >}}

Reviewer: consider moving the checkpoint after the "Below the GetImages" paragraph. also the point cloud portion is really cool, so it might be worth showing the screen shot,

The camera card shows a live color stream from `cam-1`. Because you configured both the `color` and `depth` sensors, switch the **GetImages** stream rate to "Refresh every second" and the source to depth and confirm that stream updates too. If the card is blank, check the LOGS tab for a camera error before moving on.
{{< /checkpoint >}}

Below the **GetImages** control, you can toggle **GetPointCloud** on and set the "Camera up vector" to "-y" to see a live stream of the pointcloud data from the camera as a 3D scene. Hovering over any part of the point cloud with your mouse will display the x, y, z coordinates of that point from the perspective of the camera and distance from the camera sensor.

{{< alert title="The arm is about to move" color="caution" >}}

> Jogging the arm (setting the joint sliders and pressing **Execute**) moves the physical arm. Before you run a move, confirm the workspace is clear, keep the e-stop within reach, and change one joint a small amount at a time. Large or combined joint moves can drive the arm into the table, the camera, or itself.

Reviewer: ```suggestion
The next section moves the physical arm using joint control. Before you run a move, confirm the workspace is clear, keep the e-stop within reach, and change one joint a small amount at a time. Large or combined joint moves can drive the arm into the table, the camera, or itself.
```

{{< /alert >}}

> On the arm card, set a joint with the **MoveToJointPositions** sliders and press **Execute** to move the arm. Then, under **MoveToPosition**, press **Current position** and confirm the Pose Values populate with the arm's current x, y, and z.

Reviewer: ```suggestion
The arm card shows the arm's current joint positions, and provides two different control methods to test the arm, Joint Control and Cartesian control. Joint Control lets you manually set a joint's desired position and click execute, while the cartesian control lets you move the arm to a specific pose in space. Before using either, it is always a good idea to press Current Position to set the desired values to match the robot's current state. Try moving joint 5 and 6 with the **MoveToJointPositions** sliders and press **Execute** and watch the end of the arm move. Then, under **MoveToPosition**, press **Current position** and confirm the Pose Values populate with the arm's current x, y, and z.
```

<!-- ASSET P1 control-arm-card (UI+): MoveToJointPositions sliders + Execute, and the MoveToPosition "Current position" button boxed -->

{{<imgproc src="/tutorials/pick-and-place/control-arm-card.png" resize="1200x" declaredimensions=true alt="The arm CONTROL card with joint sliders, Execute, and the MoveToPosition Current position button.">}}

{{< checkpoint >}}

> Pressing **Execute** after setting a joint slider moves the physical arm, and pressing **Current position** under **MoveToPosition** fills the Pose Values rather than erroring. If nothing moves, confirm `arm-1` shows as online in the CONFIGURE tab and that the LOGS tab has no connection errors for it.

Reviewer: ```suggestion
Pressing **Execute** after setting a joint slider moves the physical arm, and pressing **Current position** under **MoveToPosition** fills the Pose Values. If nothing moves or you see an error message, confirm `arm-1` shows as online in the CONFIGURE tab and that the LOGS tab has no connection errors for it.
```

{{< /checkpoint >}}

> On the gripper card, place a block between the gripper fingers by hand, then select **Grab**. The fingers should close and hold the block. Select **Open** and confirm the fingers release it.

Reviewer: ```suggestion
The gripper test card lets you open and close a gripper, check whether the gripper is moving and whether it believes it's holding an object. Place a block between the fingers, and press **Grab** to close the fingers and hold the block. **Open** releases the block. This grab-and-release is the same action your Python code performs later in the workshop when it picks a block and drops it in a bin. If your gripper card also shows a holding status indicator, it now reads true while the block is held and false once the gripper is open and empty.
```

<!-- ASSET P1 control-gripper-grab (MOTION): gripper closing on a block via Grab, then Open releasing -->

{{<imgproc src="/tutorials/pick-and-place/control-gripper-grab.jpeg" resize="1200x" declaredimensions=true alt="The two-finger gripper holding a block.">}}

{{< checkpoint >}}

> With a block between the fingers, **Grab** closes the fingers and the gripper holds the block without dropping it. **Open** releases the block. This grab-and-release is the same action your Python code performs later in the workshop when it picks a block and drops it in a bin. If your gripper card also shows a holding status indicator, it now reads true while the block is held and false once the gripper is open and empty.

Reviewer: ```suggestion
With a block between the fingers, **Grab** closes the fingers and the gripper holds the block without dropping it. **Open** releases the block. If nothing moves, confirm gripper-1 shows as online in the CONFIGURE tab and that the LOGS tab has no connection errors for it.
```

{{< /checkpoint >}}

## The 3D scene tab

<!-- ASSET P0 3dscene-wrist-camera (MOTION): jog joint 1 and watch the cam-1 frame move with the arm (the wrist-camera insight) -->

> Open the **3D scene** tab. You should see a rendered model of the arm and the `cam-1` frame positioned at the end of its wrist.

Reviewer: ```suggestion
Open the **3D scene** tab. The 3D scene tab uses the resources and frames you configured to build a 3D representation of the hardware. You should see a the arm, with the base at 0,0,0 in world, and a coordinate frame for `cam-1` and for the gripper positioned near the end of the wrist.
```

Jog joint 1 with the arm card's sliders and watch the 3D scene as the arm turns. The `cam-1` frame moves with the arm, because the camera is mounted on the wrist rather than fixed in the workspace.

{{< alert title="Why this matters later" color="note" >}}

> Because the camera is wrist-mounted, every detection it makes is relative to wherever the arm happens to be pointed at that moment. In Phase 5 you will detect from a single known pose so that "camera space" always means the same thing. Notice that pose here; you will meet it again as the "detect from home" rule.

Reviewer: i am not sure what it means by "single known pose"

{{< /alert >}}

> The frames you configured earlier are what make this work: because `cam-1`'s parent is `arm-1`, its frame rides along every time the arm moves. Confirm the relationship in the scene: the `cam-1` frame sits out at the wrist and tracks the arm as you jog it.

Reviewer: ```suggestion
```

## Check your work

> If you want to compare your configuration against a known-good version, the companion repo has a reference copy at [machine-fragment.json](https://github.com/viam-devrel/pick-and-place/blob/main/config/machine-fragment.json). Use it to check your work, not to import over what you just built by hand.

Reviewer: The sample has everything, so it might be good to only show the json for the 3 components here. also might want to remove " not to import over what you just built by hand."
